src/clone.rs

```diff
@@ -0,0 +1,62 @@
+use std::rc::{self, Rc};
+use std::sync::{self, Arc};
+
+pub trait Downgrade {
+    type Target;
+
+    fn downgrade(&self) -> Self::Target;
+}
+
+impl<T> Downgrade for Arc<T> {
+    type Target = sync::Weak<T>;
+
+    fn downgrade(&self) -> Self::Target {
+        Arc::downgrade(self)
+    }
+}
+
+impl<T> Downgrade for Rc<T> {
+    type Target = rc::Weak<T>;
+
+    fn downgrade(&self) -> Self::Target {
+        Rc::downgrade(self)
+    }
+}
+
+pub trait Upgrade {
+    type Target;
+
+    fn upgrade(&self) -> Option<Self::Target>;
+}
+
+impl<T> Upgrade for sync::Weak<T> {
+    type Target = Arc<T>;
+
+    fn upgrade(&self) -> Option<Self::Target> {
+        self.upgrade()
+    }
+}
+
+impl<T> Upgrade for rc::Weak<T> {
+    type Target = Rc<T>;
+
+    fn upgrade(&self) -> Option<Self::Target> {
+        self.upgrade()
+    }
+}
+
+#[macro_export]
+macro_rules! clone {
+    ($($n:ident),+ => move |$($p:pat),*| $body:expr) => (
+        {
+            $( let $n = $crate::clone::Downgrade::downgrade(&$n); )+
+            move |$($p,)*| {
+                $(let $n = match $crate::clone::Upgrade::upgrade(&$n) {
+                    Some(val) => val,
+                    None => panic!("cannot upgrade weak reference `{}`", stringify!($n)),
```

Comment:
Instead of panicking, this should probably do something like the `upgrade_weak!` macro we have in some of the examples: return with some default value if upgrading fails.

Upgrade failure is not something that should panic, it can be completely normal and expected.

Reply:
Not sure I understand how it can be normal.
If you try to access a variable that was freed, how could it not be a bug?
Do you have an example?

Reply:
Because it was freed already but e.g. your signal handler was called again nonetheless, so you just do nothing instead of exploding. This can happen during shutdown for example, when a signal handler is called right again before you disconnected it, etc.

The whole point of weak references is to be able to handle the non-existence of the underlying object gracefully.

Reply:
Do you want only one default return value even if there are multiple weak references?
I'm not sure how it could work otherwise (i.e. with weak references to widgets).

Reply:
Yes, only one. See the `upgrade_weak!` macro we have in some examples.